operator/pkg/ciliumendpointslice/controller.go

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ciliumendpointslice
+
+import (
+	"context"
+	"time"
+
+	"github.com/cilium/workerpool"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+)
+
+// params contains all the dependencies for the CiliumEndpointSlice controller.
+// They will be provided through dependency injection.
+type params struct {
+	cell.In
+
+	Logger    logrus.FieldLogger
+	Lifecycle hive.Lifecycle
+
+	Clientset           k8sClient.Clientset
+	CiliumEndpoint      resource.Resource[*v2.CiliumEndpoint]
+	CiliumEndpointSlice resource.Resource[*v2alpha1.CiliumEndpointSlice]
+
+	Cfg       Config
+	SharedCfg SharedConfig
+}
+
+type endpointEvent struct {
+	event eventType
+	cep   *v2.CiliumEndpoint
+}
+
+type Controller struct {
+	logger        logrus.FieldLogger
+	context       context.Context
+	contextCancel context.CancelFunc
+
+	// Cilium kubernetes clients to access V2 and V2alpha1 resources
+	clientset           k8sClient.Clientset
+	ciliumEndpoint      resource.Resource[*v2.CiliumEndpoint]
+	ciliumEndpointSlice resource.Resource[*v2alpha1.CiliumEndpointSlice]
+
+	// reconciler is an util used to reconcile CiliumEndpointSlice changes.
+	reconciler *reconciler
+
+	// Manager is used to create and maintain a local datastore. Manager watches for
+	// cilium endpoint changes and enqueues/dequeues the cilium endpoint changes in CES.
+	// It maintains the desired state of the CESs in dataStore
+	manager      operations
+	slicingMode  string
+	maxCEPsInCES int
+
+	// workqueue is used to sync CESs with the api-server. this will rate-limit the
+	// CES requests going to api-server, ensures a single CES will not be proccessed
+	// multiple times concurrently, and if CES is added multiple times before it
+	// can be processed, this will only be processed only once.
+	queue           workqueue.RateLimitingInterface
+	queueTerminated chan struct{}
+	writeQPSLimit   float64
+	writeQPSBurst   int
+
+	enqueuedAt                     map[CESName]time.Time
+	preInitEnqueuedEndpointsEvents []endpointEvent
+	endpointsMappingInitialized    bool
```

Comment:
I think the access to `endpointsMappingInitialized` needs to be synchronized (for example using an `atomic.Bool`). That's because it is written by the goroutine that executes `syncCESsInLocalCache` and it is read by the goroutine that executes `onEndpointUpdate` and `onEndpointDelete` (that is, the goroutine `cilium-endpoints-updater` from the workerpool).
IOW, the second goroutine is doing a busy wait on that boolean, and doing that without synchronization is incorrect, as reported [here](https://go.dev/ref/mem) (look for the relevant example regarding "busy waiting").

Reply:
I don't think it's busy waiting.

```go
func (c *Controller) onEndpointEvent(cep *cilium_api_v2.CiliumEndpoint) {
  if c.endpointsMappingInitialized {
    process
  } else {
    add to queue
}
```
Nothing else read it. It's changed to true later.

I agree we need synchronization to avoid situation in which we miss an event - check for c.endpointsMappingInitialized return false but the queue is started being processed before new element is added.

I think we need synchronization on the preInitEnqueuedEndpointsEvents even more. If two threads change it at the same time (it's not possible right now I think as we have 1 thread processing the events) then we will miss some events.

It's too bad we will have quite complicated code here just for short period of time.
I'll try to remove it all together. I think we can start processing the events only after the initial mapping is initialized.

Reply:
Done it. It is quite simple with the hive as we can start processing event when we want to.